fix: Improve text line height calculation to properly align text on iOS

ArekChr opened the pull request:
## Summary:

This pull request addresses an issue in React Native on iOS: Text gets truncated when the line height is too small, even if the line-height matches or exceeds the font size.

The existing logic did not correctly handle font metrics (ascent, descent, top, bottom) when the total height exceeded the line height. The implementation now ensures a more balanced vertical space distribution, reducing text clipping and misalignment.

The fix mirrors a similar solution made on [Android](https://github.com/facebook/react-native/pull/46362), ensuring consistency across both platforms.

For more context, see the related issue: [React Native issue #29507](https://github.com/facebook/react-native/issues/29507).

## Changelog:

[IOS] [FIXED] - Fixed an issue where text clipping and misalignment occurred when the line-height was smaller than the total ascent and descent of the font.

## Test Plan:
The result is an improvement over the current implementation, reducing the occurrence of text truncation. While minimal truncation still exists at single-line heights, particularly with special characters, the implementation better preserves the vertical alignment of the text compared to previous behavior. However, like the Android implementation, this does not fully resolve the issue of React Native truncating content that is out of bounds of the span.

| New Arch Before               | New Arch After               |
|-------------------------------|------------------------------|
| ![New Arch Before](https://github.com/user-attachments/assets/3efa8b54-0ec4-498f-83d1-8916062c847c) | ![New Arch After](https://github.com/user-attachments/assets/a56f9fb6-b619-4560-bd70-60b2b2972706) |

ArekChr committed:
improve text line height calculation on ios

mrousavy commented:
nice!

ArekChr committed:
add comments

ArekChr commented on packages/react-native/Libraries/Text/Text/RCTTextView.mm:
```diff
@@ -98,6 +99,36 @@ - (void)setTextStorage:(NSTextStorage *)textStorage
+  } else if (textHeight < lineHeight) {
+      CGFloat difference = lineHeight - textHeight;
+      verticalOffset = -(difference / 2.0);
+  }
```
Sure, added comments

ArekChr committed:
add safeguards to prevent out-of-bounds crash when accessing text attributes in calculateDrawingPoint.

ArekChr commented on packages/react-native/Libraries/Text/Text/RCTTextView.mm:
```diff
@@ -98,6 +99,36 @@ - (void)setTextStorage:(NSTextStorage *)textStorage
+                                   contentFrame:(CGRect)contentFrame {
+  UIFont *font = [textStorage attribute:NSFontAttributeName atIndex:0 effectiveRange:NULL];
+  if (!font) {
+      font = [UIFont systemFontOfSize:14];
```
I couldn’t find systemFontSize used anywhere else in the repo, and the default font size of 14 has been hardcoded. Also, since systemFontSize is not available on tvOS, I’m not sure if it would be a suitable replacement. Let me know your thoughts on whether we should stick with the hardcoded value for consistency or if there’s a preferred approach.

cipolleschi replied:
@NickGerleman 
> This seems sketch. E.g. what if there is a font multiplier applied?

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
codegen feature flags

cipolleschi reviewed (approved):
Changes looks good. We discussed this internally and we can proceed importing it.

To set the expectation, this might take some time to merge, given that this might affect internal tests and product code and we would need some time to fix them. But thankfully, we have a feature flag to control the rollout! 😄

cipolleschi commented:
@ArekChr CI is currently red. The error is
```
[React-RCTText] Compiling RCTTextView.mm
Error: use of undeclared identifier 'ReactNativeFeatureFlags'; did you mean 'facebook::react::ReactNativeFeatureFlags'?
  if (ReactNativeFeatureFlags::enableLineHeightCentering()) {
```

Could you please verify why this is happening and fix it?
You should be able to build it locally running these commands in the react-native root folder:
```
yarn install
cd packages/rn-tester
bundle install
bundle exec pod install
open RNTesterPods.xcworkspace
```
and then build from Xcode.

If you need to run the old architecture, you can disable it by adding:

```objc
- (BOOL) newArchEnabled
{
  return NO;
}
```
in the AppDelegate.mm file

ArekChr committed:
Fix namespace for ReactNativeFeatureFlags in RCTTextView

ArekChr commented:
Hi @cipolleschi, the issue has been fixed, and the build runs successfully locally. Thanks!

cipolleschi commented:
We were discussing internally that it would be better to have two separate feature flags, one for Android and one for iOS. Could I ask you to:
1. revert the changes you made on the preexisting feature flag, which was targeting only Android
2. add the iOS dual of that feature flag

The idea is to be able to control the two platforms separatedly.

Thank you so much for your help and patience

ArekChr committed:
refactor: separate feature flat per platform

ArekChr committed:
fix: add React-featureflags pod for dynamic frameworks configuration

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr commented:
Hi, @cipolleschi. The CI failed due to some issues unrelated to my changes. I’ve likely fixed the problem with dynamic frameworks, but two workflows seem missing. Could you please trigger those workflows again so we can proceed?

cipolleschi commented:
could you rebase on top of main, please? 🙏

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
update feature flags

ArekChr committed:
fix: feature flag import in rn-tester dynamic frameworks

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
fix: update feature flags

ArekChr committed:
Merge remote-tracking branch 'upstream/main' into improve-lineheight-calc-ios

Blyawon commented:
Are there any updates on this?

cipolleschi commented:
@ArekChr could you rebase on main?

ArekChr committed:
Merge remote-tracking branch 'upstream/main' into improve-lineheight-calc-ios

ArekChr commented:
@cipolleschi sure, done!

cipolleschi commented:
JS tests and the linter are failing. :( Could you have a look at those failures, please?

ArekChr committed:
fix: add missing expectedReleaseValue to feature flags config

ArekChr commented:
@cipolleschi Fixed

ArekChr committed:
refactor: remove unused flag on android

ArekChr committed:
Merge remote-tracking branch 'upstream/main' into improve-lineheight-calc-ios

LJNGDAHL commented:
Any progress on this? 👀

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

cipolleschi commented:
Sorry if it is taking a bit. I'm trying to get an internal review on this.

ludvigbartholdsson commented:
Awesome if we could get this PR merged (if it works and passes tests ofc) @cipolleschi do you think it's possible next week?

0xbenon commented:
I hope you're well. I wanted to check on the status of the lineHeight fix, as it’s quite important for my project. Thank you for your efforts

ludvigbartholdsson commented:
Hi @cipolleschi , do you have a new status on this?

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

cipolleschi commented on packages/react-native/Libraries/Text/Text/RCTTextView.mm:
```diff
@@ -98,6 +99,42 @@ - (void)setTextStorage:(NSTextStorage *)textStorage
   [self setNeedsDisplay];
 }
 
+- (CGPoint)calculateDrawingPointWithTextStorage:(NSTextStorage *)textStorage
```
@NickGerleman 
> It would be best to keep the changes to the Fabric component instead of this one.

ArekChr replied:
Could you point me to the specific place in the Fabric component where I should move this change?

SimpleCreations replied:
Hi! I'm a developer following this thread. Does this suggestion mean that this fix will not be made for the old architecture? I've been following the original ticket for years, and such outcome would be quite disappointing 😕 My team is not considering upgrading to Fabric any time soon (due to many old arch dependencies, and also performance/bundle size/reliability concerns).

cipolleschi replied:
@ArekChr I believe that the suggestion here was to not touch the files in the old architecture. The PR already updates the New Architecture, so we are good.
---
@SimpleCreations Thanks for the feedback, I believe that in this case we can keep the fix here. As a general thought, though.. Yes, in the future the Old Architecture will not be maintained. 

Right now, we already released 0.76 and 0.77 where the New Architecture is the default. We are already working on 0.78 and once that's out, 0.75 will go out of support. That is the last version where the Old Architecture is used by default.

Consider that new features are only developed for the New Architecture and many libraries will be only compatible with the New Architecture, moving forward (reanimated and react-native-vision-camera to mention two of them). So we strongly advise to start migrating to the New Architecture.

I'd be curious to know what is holding you back in more details. If you can replicate Performance and bundle issues in separate reproducers, we are more than happy to look into them as soon as possible.

SimpleCreations replied:
@cipolleschi Thank you for reconsidering! I appreciate the decision to keep the fix very much.

I'll try to answer your question about what is holding my team back. The primary reason is we use quite a few SaaS solutions that provide their own React Native SDKs that are old arch only, and they have no timeline of adding new arch support. Also we have developed a few of our own React Native wrappers for other SaaS solutions, and they are shared between other apps in our company, which are also on old arch.

Performance and bundle size concerns come from me following this thread: https://github.com/facebook/react-native/issues/47490 where multiple people share negative experiences with new arch migration, but I'm not sure if this would affect us, because we haven't tried migrating yet (due to what I described above). So please disregard this part for now 🙂

I would like to also mention that the same fix for line height has been made for old arch on Android here: https://github.com/facebook/react-native/pull/46362, so it would make sense to include an old arch fix on iOS too.

cipolleschi commented on packages/react-native/Libraries/Text/Text/RCTTextView.mm:
```diff
@@ -98,6 +99,42 @@ - (void)setTextStorage:(NSTextStorage *)textStorage
+    return contentFrame.origin;
+  }
+
+  UIFont *font = [textStorage attribute:NSFontAttributeName atIndex:0 effectiveRange:NULL];
```
@NickGerleman 
> Do you know if this is influenced by the full set of text attributes on the Text component? Like, if this is italicized, will it pick up that version of the font?

ArekChr replied:
This code will return the correct font if the entire text is consistently styled (e.g., italic or bold). However, if the text contains different styles in different parts (e.g., a mix of italic and bold), this code will only reflect the style of the first character.

cipolleschi commented on packages/react-native/Libraries/Text/Text/RCTTextView.mm:
```diff
@@ -98,6 +99,42 @@ - (void)setTextStorage:(NSTextStorage *)textStorage
+  CGFloat verticalOffset = 0;
+  // Adjust vertical offset to ensure text is vertically centered relative to the line height.
+  // Positive offset when text height exceeds line height, negative when line height exceeds text height.
+  if (textHeight > lineHeight) {
+      CGFloat difference = textHeight - lineHeight;
+      verticalOffset = difference / 2.0;
+  } else if (textHeight < lineHeight) {
+      CGFloat difference = lineHeight - textHeight;
+      verticalOffset = -(difference / 2.0);
+  }
```
@javache 
There's no need for the comparison.
```suggestion
  CGFloat difference = textHeight - lineHeight;
  CGFloat verticalOffset = difference / 2.0;
```

ArekChr replied:
I’ve updated the fallback to use UIFontMetrics so it respects dynamic type scaling

cipolleschi commented on packages/react-native/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm:
```diff
@@ -326,6 +327,40 @@ @implementation RCTParagraphTextView {
+  CGFloat lineHeight = font.lineHeight;
+
+  if (paragraphStyle && paragraphStyle.minimumLineHeight > 0) {
+      lineHeight = paragraphStyle.minimumLineHeight;
```
@NickGerleman 
> Why do we do this work when an explicit lineHeight is not set?

ArekChr replied:
I’ve inverted the condition to make it clearer and ensure we exit early when minimumLineHeight isn’t set.

cipolleschi commented on packages/react-native/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentView.mm:
```diff
@@ -326,6 +327,40 @@ @implementation RCTParagraphTextView {
+  CGFloat verticalOffset = 0;
+  // Adjust vertical offset to ensure text is vertically centered relative to the line height.
+  // Positive offset when text height exceeds line height, negative when line height exceeds text height.
+  if (textHeight > lineHeight) {
+      CGFloat difference = textHeight - lineHeight;
+      verticalOffset = difference / 2.0;
+  } else if (textHeight < lineHeight) {
+      CGFloat difference = lineHeight - textHeight;
+      verticalOffset = -(difference / 2.0);
+  }
```
@NickGerleman 
> nit: these checks are redundant, since one inverts both the subtraction order, and then inverts again for the difference.

```suggestion
  CGFloat difference = textHeight - lineHeight;
  CGFloat verticalOffset = difference / 2.0;
```

cipolleschi commented:
I am sorry that this is taking so long, but this diff is very delicate to land as it touches the very core of text rendering and it can introduce a lot of regression. That's why it is take more time. 
On top of that, there is further feedback that needs to be addressed and that i forwarded from internal colleagues.

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
simplify offset calculation

ArekChr committed:
Handle font multiplier for dynamic type in text storage fallback

ArekChr committed:
Exit early if minimumLineHeight is not set

cipolleschi commented:
need a rebase due to the feature flags. 😢

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
update feature flags

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

Perronef5 commented:
WEN Merge

ArekChr committed:
fix positioning when lineheight > textheight

ArekChr committed:
fix: font metrics on ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

SimpleCreations commented:
@ArekChr amazing work! I have been having issues with custom font ascenders/descenders, and I'm so glad to see there's an effort to make it consistent across Android and iOS.

One concern I have is that you mentioned that iOS behavior is made to match Android. However, I have a font that is aligned more desirably on iOS and web, and is misaligned on Android, and I'm not sure if it's the problem is in the font file itself, or if this means that the Android implementation is lacking. I can only share the font file in private though (it is not free)

cipolleschi commented:
this is really going out of my specialization area. I'll try to pass the PR over to the people that are taking care of working on text improvements. 

I want to stress that this change is a very delicate one: this changes might affect the whole product code in all the apps, bith the ones we use internally at Meta and the ones developed by the community, so we need to take our time to make sure that it does not break all the apps written with React Native.

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr commented:
Hey @cipolleschi, just checking in. Do you have any update on this?

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

cipolleschi commented:
@ArekChr I opened a ticket internally to find a better person to handle this, as it is out of my area of expertise.
The general sentiment is that it is something we do want, but the problem is that we don't have free capacity to work on this.

Even if we land the change as it is, behind feature flags so with no effect on the current behavior, we still need a Meta engineer that can run experiment and, eventually, enable the feature flag for everyone.

These feature flags are not something that our final users are supposed to change: so even by landing this, user would not probably benefit from it until we enable it for everyone.

I'll keep poking at people, asking to prioritize this work, but it might take a bit. :(

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
remove old arch impl

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Update RCTParagraphComponentView.mm

Co-authored-by: troZee <12766071+troZee@users.noreply.github.com>

thymikee commented:
@cipolleschi any chance we can push this forward? It’s a patch we need to apply in a few projects

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
fix: run featureflags gen

ArekChr committed:
sync pods in rn-tester

ArekChr committed:
fix(iOS): properly modify frame for line height centering

- Rename calculateCenteredFrameWithAttributedText to adjustFrameForLineHeightCentering
- Change method to accept CGRect pointer instead of passing by value
- Properly modify frame using pointer dereferencing (frame->origin.y)

ArekChr committed:
fix(iOS): align line height centering rounding with Android

Changed from ceil/floor to round with remainder calculation to
prevent 1-pixel vertical shifts and ensure cross-platform parity.

ArekChr commented:
Rebased with code style improvements. This PR fixes the long-standing iOS/Android line height centering discrepancy. Screenshots showing platform parity attached.

| Before | After |
|--------|-------|
| <img width="420" alt="before-repro" src="https://github.com/user-attachments/assets/898b1338-0484-48c7-98c4-ccd8deb193a4" /> | <img width="420" alt="after-repro" src="https://github.com/user-attachments/assets/7256b1c2-9f36-4ca6-8115-433e38878aed" /> |
| <img width="420" alt="before-example" src="https://github.com/user-attachments/assets/56ed10c3-8b64-4e29-9d4f-6fca85d81935" /> | <img width="420" alt="after-example" src="https://github.com/user-attachments/assets/eb6a9bfd-8d4c-4279-85b1-fd4de9fe834f" /> |

DimitarNestorov commented:
>  would this also resolve the issues ... on Android

@matt-oakes as per the PR title this fix is for iOS only

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
Merge branch 'main' into improve-lineheight-calc-ios

ArekChr committed:
remove feature flag

ArekChr committed:
Revert "sync pods in rn-tester"

This reverts commit c7407afaa60d3d0b27280d2e8575452d1e0500b5.
